src/elm/external.ts

```diff
@@ -22,21 +25,49 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype, this);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype
+    };
+
+    let resolvedCodes: Code[] | ValueSet | undefined;
+    if (this.codes) {
+      resolvedCodes = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
-    }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
+
+      retrieveDetails.codes = resolvedCodes;
     }
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
+    let range: Interval | undefined;
     if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+      range = this.dateRange.execute(ctx);
+
+      retrieveDetails.dateRange = range;
+      retrieveDetails.dateProperty = this.dateProperty;
```

Comment:
And as noted above, I think we should avoid making assumptions about what properties are or are not useful in combination with other properties.  So I think that if `dateProperty` exists, even with no `dateRange`, we should go ahead and pass it through.  I can't imagine how it would be used, but there's no reason for us to suppress information that is in the ELM.

In other words, I suggest moving the assignment of `retrieveDetails.dateProperty` outside of this block.